### ISSUE TITLE
Getting started doc please

### ISSUE THREAD

**Issue by suntong** [cogentcore/core#1108]
At least for _"Getting started to use core for web app development"_

Hi, I'd like to try out cogentcore immediately after I read it, but having gone through the blog and every related section of the doc in https://www.cogentcore.org/core/, like https://www.cogentcore.org/core/advanced/building, I realized that I'm still have only a vague idea how to use core for web app development. 

For how to get started with even the two tutorial demos, I realize that I have to read between the lines to make it work at my end. Thus, I'm hoping to have a clear Getting started doc for developing web apps (under Linux) so that I (or the next guy) don't need to guess.

Thanks for considering

**Comment by kkoreilly** [comment]
Thank you for the feedback. I'm not completely sure what you're asking for, but I would be glad to improve the documentation if you have a specific suggestion. As documented [here](https://www.cogentcore.org/core/setup/building), you can run an app on the web using `core run web`. So, to get started with web app development, you could try copying the [hello world](https://www.cogentcore.org/core/basics/hello-world) example and running it using `core run web`, and then progress to the tutorials as you mentioned. If the tutorials don't make sense, you can watch our videos explaining them on our [YouTube channel](https://www.youtube.com/@CogentCore). Again, I am happy to improve the documentation if you have any suggestions you think would be helpful.

**Comment by suntong** [comment]
Thanks, this is good enough for me to get started. 

BTW, the install doc is at https://www.cogentcore.org/core/setup/install

Oh, and after I do `core run web`, where to check my web site? I remember read it somewhere, like localhost:8080, but I just cannot find it any more...

```sh
$ curl localhost:8080
curl: (7) Failed to connect to localhost port 8080 after 0 ms: Couldn't connect to server
```

**Comment by suntong** [comment]
after I do core run web, where to check my web site pls?

**Comment by kkoreilly** [comment]
Yes, it is http://localhost:8080. It prints that when you do `core run web`; you need to keep the `core run web` command running for it to keep serving (or you can just do `core build web` and then pull up the `index.html` file directly if you want) (you can also use `core run web &` to have the command keep running in the background).

**Comment by suntong** [comment]
> It prints that when you do `core run web`

Oh, hmm, it prints nothing for me. It's about five minutes now. 

My core setup was successful:

```
Setting up libglvnd-dev:amd64 (1.7.0-1build1) ...
Setting up xorg-dev (1:7.7+23ubuntu3) ...
Setting up libgl1-mesa-dev:amd64 (24.0.9-0ubuntu0.1) ...
. . .
No VM guests are running outdated hypervisor (qemu) binaries on this host.
core setup succeeded
```

hmm, I remember seeing Linux installation issue somewhere but apparently not from https://www.cogentcore.org/core/setup/install. This is the problem, so many things scattered around, so even the Getting started doc only links to relevant pages would be helpful.

BTW, I found the GH issue from my history, and found where the problem is:

```
$ vulkaninfo
Command 'vulkaninfo' not found, but can be installed with:
sudo apt install vulkan-tools
```

I.e, my core setup _was successful_, but apparently _not enough_.

Reopen for doc enhancement...

**Comment by kkoreilly** [comment]
Yes, I am improving the docs now to more clearly link to the right installation steps. I will respond to your other issue (#1109) now.

**Comment by suntong** [comment]
> I am improving the docs now to more clearly link to the right installation steps

Great. Thanks. 

If you have a dedicated Getting started doc or anything, would you add in how to change page title and site icon as well please? Thanks

**Comment by kkoreilly** [comment]
Yes, I will add docs for that now.

**Comment by suntong** [comment]
> `Serving at http://localhost:8080`

Oh, BTW, please clarify if it only servers for localhost _in the add docs_, and how to open up to the world if not. thx.

**Comment by kkoreilly** [comment]
I added docs for all of those things in #1113. Here is the relevant information:

> The first call in every Cogent Core app is [[core.NewBody]]. This creates and returns a new [[core.Body]], which is a container in which app content is placed. This takes an optional name, which is used for the title of the app/window/tab.

> Note: `core run web` is mainly used for serving an app locally; to deploy an app on the web, you can use `core build web` and a static site host (here is an [example](https://github.com/cogentcore/cogentcore.github.io/blob/main/.github/workflows/core.yml) using GitHub Pages).

> If there is an `icon.svg` file in the current directory, it will be used as the app icon on all platforms.

**Comment by suntong** [comment]
_Re using the old thread so all relevant info are in the same place..._

I've checked

- https://www.cogentcore.org/core/widgets/other/pages
- https://pkg.go.dev/cogentcore.org/core/pages, and
- https://github.com/cogentcore/core/blob/main/pages/examples/basic/basic.go

and found them not enough to be used to start building _multi-page_ web applications. 

If I have a web flow like the following, 

```
welcome
`--main
  `--page1
  `--page2
    `--page2-subp1
    `--page2-subp2
  `--page3
```

How can I structure my welcome.go and main.go, if I can modulize them separately, and main.go will call page2.go, which contains for itself and also its child pages.

Basically, it's an example of 

- being able to provide a welcome page that can goes nowhere but to main
- being able to build separate things separately, and can also 
- being able to group different level of simple things together in a same file. 

Please help. thanks

**Comment by kkoreilly** [comment]
Are you effectively trying to reproduce the same nested directory structure as `pages` but with Go files instead of markdown files?

**Comment by suntong** [comment]
No, I'm not trying to reproduce the same nested directory structure as pages but with Go files instead of markdown files. Everything works differently, I just want to see how cogentcore best handles such _multi-page multi-level web applications_. Moreover, neither of the pages should be limited to markdown files, like page2 can be an input form and page2-subpX are just the outcome of the submit. In fact, please do so if possible.

Overall, _**viewing it as black-box from the outside**_, either from the client point of view or files/dirs arrangement point of view, it should

> * being able to provide a welcome page that can goes nowhere but to main
> * being able to build separate things separately (if possible), and can also
> * being able to group different level of simple things together in a same file.

I just want to see how cogentcore best handles such cases, so that I can start mine. 

Thanks

**Comment by theclapp** [comment]
My impression is that @suntong wants to know how to completely replace most or all of the widgets on a page, programatically.

For example, the examples/demo window has a top menu-bar (not sure what to call it), and a tab bar with five tabs. Imagine if instead you wanted no menu-bar-thingy, and just the widgets on the first tab, and a "next" button, which showed everything in the "widgets" tab, with a "prev" and "next" button, and that "next" button showed everything in the "collections" tab, with a "prev" and a "next button, etc.

I think (but am not sure) that you can arrange this with a "Maker" function for the toplevel widget, which creates a "plan" with entirely different sets of widgets in it, and the plan changes based on some top-level "stage" or "page id" variable.

The https://www.cogentcore.org/core code may hint at how to do this by looking at how it replaces the main content on the right after you click on a topic in the tree on the left. Ditto the implementation of the Tabs widget, and probably some others.

**Comment by suntong** [comment]
Yeah, those are kind of implementation details, which I have no restriction/limitation. Just think of it as a login page, the normal look would be a floating window on top of a blank page. Or, if it's not possible, it can be as simple as --

https://suntong.github.io/demo-todo/

where the login name input is used in the next main page.

**Comment by kkoreilly** [comment]
Yes, thank you for clarifying, I understand what you are trying to do now. I will give a more detailed response in a little bit.

**Comment by kkoreilly** [comment]
When we do this in our apps, we mainly use full-window dialogs for new pages. So, for example, if you have a login page as your main page, then once someone presses the login button, it creates a full-window dialog for the next page. For example, something like this:

```go
func main() {
	b := core.NewBody()
	bt := core.NewButton(b).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		home := core.NewBody("Home")
		core.NewText(home).SetText("Page content")
		home.RunFullDialog(bt)
	})
	b.RunMainWindow()
}
```

The user can then use the back button and search menu to navigate between pages if they need to. You can also move the content of each page to separate functions/files/packages to make the code nicer as you mentioned above, such as by adding a `makeHome` function:

```go
func main() {
	b := core.NewBody()
	bt := core.NewButton(b).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		makeHome(bt)
	})
	b.RunMainWindow()
}

func makeHome(ctx core.Widget) {
	home := core.NewBody("Home")
	core.NewText(home).SetText("Page content")
	home.RunFullDialog(ctx)
}
```

If this approach does not satisfy your needs, I would be open to adding infrastructure for another approach, such as one based on plans as @theclapp suggested above.

**Comment by suntong** [comment]
Thanks @kkoreilly, 

Suppose the welcome page takes a username as input, how to use/show its value in the next main page pls? thx.

**Comment by kkoreilly** [comment]
You can add a text field and then pass its value to the next page:

```go
func main() {
	b := core.NewBody()
	username := core.NewTextField(b)
	bt := core.NewButton(b).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		makeHome(bt, username.Text())
	})
	b.RunMainWindow()
}

func makeHome(ctx core.Widget, username string) {
	home := core.NewBody().AddTitle("Hello, " + username)
	core.NewText(home).SetText("Page content")
	home.RunFullDialog(ctx)
}
```

Alternatively, you could make a `user` struct and display it using a `core.Form`, which is more extensible such as if you want to add more fields:

```go
type user struct {
	Username string
}

func main() {
	b := core.NewBody()
	u := &user{}
	core.NewForm(b).SetStruct(u)
	bt := core.NewButton(b).SetText("Sign in")
	bt.OnClick(func(e events.Event) {
		makeHome(bt, u)
	})
	b.RunMainWindow()
}

func makeHome(ctx core.Widget, u *user) {
	home := core.NewBody().AddTitle("Hello, " + u.Username)
	core.NewText(home).SetText("Page content")
	home.RunFullDialog(ctx)
}
```

**Comment by kkoreilly** [comment]
For reference, our new `core.Pages` feature is the best way to achieve this multi-page structure now (see the [sign in tutorial](https://www.cogentcore.org/core/tutorials/sign-in), #1293, #1289).
